### ISSUE TITLE
Conda is installing unrequested feature

### ISSUE THREAD

**Issue by tadeu** [conda/conda#2384]
This seems to be a regression in `conda` 4.0.

I have created two test packages just to check this. They are empty and have just the `meta.yaml`:

test-release:

``` yaml
package:
  name: test
  version: "1.0"

build:
  number: 0
  features:
```

test-debug:

``` yaml
package:
  name: test
  version: "1.0"

build:
  number: 1
  features:
    - debug

requirements:
  run:
    - debug
```

(`debug` is the anaconda metapackage that only tracks the `debug` feature.)

Now, when creating a new environment with only the `test` package, it tries to install the `debug` version:

``` sh
$ conda create -n test-pkg test=1.0
Fetching package metadata: ......
Solving package specifications: .........

Package plan for installation in environment D:\Miniconda\envs\test-pkg:

The following NEW packages will be INSTALLED:

    debug: 1.0-0
    test:  1.0-debug_1 [debug]

Proceed ([y]/n)?
```

**Comment by mcg1969** [comment]
What happens if you use the same build number for both?

**Comment by tadeu** [comment]
Then it won't try to install `debug`. It only happens when the build number of the package with the feature is higher.

**Comment by mcg1969** [comment]
Quick question: you say 4.0, but are you using the latest version  (4.0.5), or truly just 4.0?

**Comment by tadeu** [comment]
Sorry, it is 4.0.5. I just assumed the regression was on the change from 3 to 4, since it seems the dependency solving was overhauled.

**Comment by tadeu** [comment]
Any news on this?
We are having to work on a "brittle" version of conda here --'.
Thanks!

**Comment by SylvainCorlay** [comment]
I think that this might be related to #4624.

**Comment by mcg1969** [comment]
Regarding this one, I think I now understand what's going on.

When performing the command `conda create -n test-pkg test=1.0`, conda's _first_ priority is to obtain the latest version and build of the packages specified on the command line. In this case, that means it is going to favor `test-1.0-1`. Since the only version of the package with this version and build number is the debug version, it _must_ install that one. 

The bug here is that features are never supposed to be pulled in involuntarily. So conda shouldn't be allowed to select the feature in that case.

Now, if you give both packages the _same_ build number, `conda create -n test-pkg test=1.0` now has a choice. There are two packages with the same version and build number, one with the `debug` feature, and one without. In this case, conda _always_ picks the one with the fewest features. So the non-debug version wins. Again, that's correct behavior.

If you want to force the debug feature in that second case, you must do `conda create -n test-pkg test=1.0 debug`. That will guarantee that the `debug` version of the package is installed. 

So the lesson here is that if you want to avoid strange behavior, the build numbers need to line up.

However, I can understand why that would be confusing. Frankly, features _are_ confusing, and brittle, and I would prefer we not use them, and find other ways to implement our requirements.

**Comment by mcg1969** [comment]
Agreed, @SylvainCorlay, this may indeed be related. Do you use the exact same build numbers for your `cling` builds?

**Comment by SylvainCorlay** [comment]
Yes I did! Thanks for looking into this.

**Comment by kalefranz** [comment]
Looks like we have a resolution and understanding at least if this particular issue.

**Comment by mcg1969** [comment]
Well, if @SylvainCorlay is using the same build numbers there's still something else going on. I'm not sure that either this issue or his other issue should be closed, because neither has really been solved.

**Comment by mcg1969** [comment]
OK, #4647 should address the _specific_ problem here. I am not convinced this will fix @SylvainCorlay's issue however.

**Comment by SylvainCorlay** [comment]
@mcg1969 applying your patch locally (I first tried or 2min installing conda from source without success)

the randomness of the behavior is gone, but it is consistently wrong: I always get the version of the package that has the `default` feature while I am installing the meta package that tracks the `cling` feature.

**Comment by SylvainCorlay** [comment]
Nevermind, still some random behavior.... Working on a reproducible example.

**Comment by mcg1969** [comment]
Yeah, I didn't expect that the patch would solve your particular problem. I'm convinced that conda is not picking up the `feature` metadata in the `cling` versions of the package. Whether that's because the metadata is not there, or conda is failing to see it, I don't know. But that would explain the randomness---because if the `feature` metadata is missing, then both versions of the package are equally valid.

**Comment by mcg1969** [comment]
Feel free to email me the `cling-patches` metapackage and one each of the `cling`/non-`cling` tarballs and I can inspect them to see if anything is amiss.

**Comment by SylvainCorlay** [comment]
- The `cling-patches` meta-package is on `conda-forge`. It tracks the `cling` feature.
 - Then we have `llvmdev` and `clangdev` also on conda-forge.

**Comment by mcg1969** [comment]
That's grea, thanks!

**Comment by mcg1969** [comment]
OK, there is definitely something wrong here:

- This [cling-patches package](https://anaconda.org/conda-forge/cling-patches/files/modal/info/58947a59966a5a10fb536055) does _not_ have `track_features: cling`.
- This [llmdev package](https://anaconda.org/conda-forge/llvmdev/files/modal/info/589f1281f9e4b2115793e1fd) _does_ have `track_features: cling`, and it _depends_ on `cling-patches`.

This definitely isn't what I expected to see, nor do I expect you. These packages have not been built to your expectations, it seems.

**Comment by mcg1969** [comment]
There are quite a few other features/track_features problems with these builds. They're probably going to need to be removed from the channel and fixed.

**Comment by mcg1969** [comment]
I stand corrected, `cling-patches` is built correctly. It's `llvmdev` and `clingdev` that are not built correctly. They shouldn't _depend_ on `cling-patches`, they need `features: cling`.

**Comment by SylvainCorlay** [comment]
Yes, indeed, I eventually removed the `feature: cling` for llvm-dev.

**Comment by mcg1969** [comment]
OK, once `llvmdev` and `clingdev` have been cleaned up I would expect the behavior to normalize. In particular, `cling-patches` must be removed from their dependencies, and the `features` list needs to include `cling`. It seems that `llvmcling` and `clangcling` are leftovers from previous attempts, so they should be removed as well. In short, everything should be alinged on the `cling` feature, and _no_ packages should depend on `cling-patches` at all.

**Comment by SylvainCorlay** [comment]
I have cleaned up the recipes locally and could check that everything works
as expected! Woot!

Things are being uploaded on the `cling-testing` channel for testing. (At
super-slow speed from a remote country-side location).

Thanks a lot for taking the time to look at this. As soon as someone in the
conda-forge team can delete the faulty builds from the channel, I will
update it.

FYI, the reason why I initially depended on the cling-patches meta package
was because of a failure to install at testing time. In fact I still need
the cling-patches package as a testing requirement for the cling variants.

Finally, clang[cling] requires llvm[cling] at build time. For the right
flavor of llvm to be picked up I had to make cling-patches a build
requirement of clang[cling].

Sylvain

On Feb 18, 2017 3:21 PM, "Michael C. Grant" <notifications@github.com>
wrote:

OK, once llvmdev and clingdev have been cleaned up I would expect the
behavior to normalize. In particular, cling-patches must be removed from
their dependencies, and the features list needs to include cling. It seems
that llvmcling and clangcling are leftovers from previous attempts, so they
should be removed as well. In short, everything should be alinged on the
cling feature, and *no* packages should depend on cling-patches at all.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/conda/conda/issues/2384#issuecomment-280848571>, or mute
the thread
